LiS/man/man4/sad.4.man
'\" rtp
.\" -*- nroff -*- vim: ft=nroff nocin nosi
.\"
.\" @(#) $Id: sad.4.man,v 1.1.4.4 2008-04-29 08:33:19 brian Exp $
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.  The author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.\" Last Modified $Date: 2008-04-29 08:33:19 $ by $Author: brian $
.\"
.\" -------------------------------------------------------------------------
.\"
.\" $Log: sad.4.man,v $
.\" Revision 1.1.4.4  2008-04-29 08:33:19  brian
.\" - update headers for Affero release
.\"
.\" Revision 1.1.4.3  2006/04/05 09:42:35  brian
.\" - moved .so line and updated headers
.\"
.\" =========================================================================
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database lis.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
.R2
.so lis.macros
.\"
.\"
.TH SAD 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B sad
\- \fISTREAMS\fP Administrative Driver
.\"
.\"
.SH SYNOPSIS
.B #include <sys/types.h>
.br
.B #include <sys/stropts.h>
.br
.B #include <sys/sad.h>
.HP
.BI "int ioctl(int " fd ", int " cmd ", int" arg );
.\"
.\"
.SH DESCRIPTION
.PP
.I fd
is a file descriptor returned from an
.BR open (2)
system call on one of the following devices:
.TP
.RI \*(lq /dev/sad/admin \*(rq
This device is only accessible by privileged users.
.TP
.RI \*(lq /dev/sad/user \*(rq
This device is accessible by all users.
.\"
.\"
.SH IOCTLS
.PP
The
.B sad
driver implements the IO controls that can be performed on an open device
using the
.BR ioctl (2)
system call as follows:
.HP
.BI "int ioctl(int " fd ", int " cmd ", int" arg );
.TP
.I fd
is an open file descriptor for the
.RI \*(lq /dev/sad/admin \*(rq
or
.RI \*(lq /dev/sad/user \*(rq
device.
.TP
.I cmd
is the ioctl command and can be one of the following:
.sp
.TS
tab(;);
lbw(12)lsss.
SAD_GAP;get autopush list.
SAD_SAP;set autopush list or clone minor.
SAD_VML;verify module list.
.TE
.TP
.I arg
is a pointer to a
.B strapush
or
.B str_list
structure, depending on
.IR cmd .
.PP
Each
.I cmd
is described in the subsections which follow:
.\"
.\"
.SS SAD_SAP
.PP
Set the autopush list for a
.IR STREAMS (4)
driver.
.PP
.I arg
points to a
.B strapush
structure formatted as follows:
.RS
.nf
\fC\s-1
struct strapush {
        unsigned int sap_cmd;
        long sap_major;
        long sap_minor;
        long sap_lastminor;
        long sap_npush;
        char sap_list[MAXAPUSH][FMNAMSZ + 1];
};
\s+1\fP
.fi
.RE
.TP \w'sap_lastminor\(em'u
.I sap_cmd
Specifies the command to execute when setting the autopush list for a
.IR STREAMS (4)
driver.
.I sap_cmd
can be one of the following values:
.RS
.TP \w'SAP_CLEAR\(em'u
.B SAP_CLEAR
Requests that the autopush list for the
.IR STREAMS (4)
driver specified by the
.I sap_major
major device number be cleared.
.TP
.B SAP_ONE
Requests that the autopush list for the
.IR STREAMS (4)
driver specified by the
.I sap_major
and
.I sap_minor
device number be set to the modules named in the
.I sap_npush
and
.I sap_list
members.
.TP
.B SAP_RANGE
Requests that the autopush list for the
.IR STREAMS (4)
driver specified by the
.I sap_major
and range of minor devices from
.I sap_minor
to
.I sap_lastminor
(inclusive) be set to the modules named in the
.I sap_npush
and
.I sap_list
members.
.TP
.B SAP_ALL
Requests that the autopush list for all minor devices of the
.IR STREAMS (4)
driver specified by the
.I sap_major
major device number be set to the modules named in the
.I sap_npush
and
.I sap_list
members.
.TP
.B SAP_CLONE
Requests that the minor device of the
.I STREAMS
driver specified by the
.I sap_major
and
.I sap_minor
device number to set to be a clone minor device.
.RE
.TP
.I sap_major
Specifies the major device number of the
.IR STREAMS (4)
driver upon which to perform
the command specified in
.IR sap_cmd .
.I sap_major
is significant for all autopush commands.
.TP
.I sap_minor
Specifies the minor device number, or initial minor device number in a range,
of the
.IR STREAMS (4)
driver for which to execute the autopush command.
.I sap_minor
is only significant for autopush commands
.BR SAP_CLEAR ", " SAP_ONE " and " SAP_RANGE ,
and is ignored otherwise.
.TP
.I sap_lastminor
Specifies the minor device number of the last minor device in a range for the
.IR STREAMS (4)
driver for which to execute the autopush command.
.I sap_lastminor
is only significant for autopush commands
.BR SAP_RANGE ,
and is ignored otherwise.
.TP
.I sap_npush
Specifies the number of module names in the
.I sap_list
member array.
.I sap_npush
is only significant for autopush commands
.I SAP_ONE ", " SAP_RANGE " and " SAP_ALL ,
and is ignored otherwise.
.TP
.I sap_list
Specifies an array containing
.I sap_npush
names of modules.
.PP
When calling
.BR SAD_SAP ,
the user must provide sufficient room for the entire 
.B strapush
structure, regardless of the number of module names in the
.I sap_list
member.
.\"
.\"
.PP
Upon success,
.B SAD_SAP
returns zero (0) and sets or clears the autopush list for the specified
.IR STREAMS (4)
device.
Upon failure,
.B SAD_SAP
returns -1, does not alter autopush list for any
.IR STREAMS (4)
device, and sets
.IR errno (3)
to an appropriate error number as follows:
.TP \w'[EEXIST]\(em'u
.RB [ EFAULT ]
.I arg
is
.B NULL
or points outside the caller's valid address space.
.TP
.RB [ EINVAL ]
.IR sap_major ", " sap_minor " or " sap_lastminor
was invalid.
.TP
.RB [ EINVAL ]
.I sap_nlist
is less than one or greater than
.BR MAXAPUSH .
.TP
.RB [ EINVAL ]
.I sap_list
contained invalid information (e.g. the names are not null terminated).
.TP
.RB [ EHOSTR ]
.IR sap_major ", " sap_minor " and " sap_lastminor
specify a character device that is not a
.IR STREAMS (4)
device.
.TP
.RB [ EEXIST ]
.IR sap_major ", " sap_minor " and " sap_lastminor
specify a
.IR STREAMS (4)
device that is already configured for autopush and the
.I sap_cmd
was
.BR SAP_ONE ", " SAP_RANGE " or " SAP_ALL .
.TP
.RB [ ERANGE ]
.I sap_minor
was not the first minor an a range previously set with
.B SAP_RANGE
and the
.I sap_cmd
was
.BR SAP_CLEAR .
.TP
.RB [ ERANGE ]
.I sap_lastminor
is less than or equal to
.I sap_minor
and the
.I sap_cmd
was
.BR SAP_RANGE .
.TP
.RB [ ENODEV ]
the device specified by
.IR sap_major " and " sap_minor
is not configured with an autopush list and the
.I sap_cmd
was
.BR SAP_CLEAR .
.TP
.RB [ ENSOR ]
.IR STREAMS (4)
resources could not be allocated to complete the command.
.PP
In addition, any of the errors returned by the
.BR ioctl (2)
command may be returned.
.\"
.\"
.SS SAD_GAP
.PP
Get the autopush list for a
.IR STREAMS (4)
driver.
.PP
.I arg
points to a
.B strapush
structure formatted as follows:
.RS
.nf
\fC\s-1
struct strapush {
        unsigned int sap_cmd;
        major_t sap_major;
        minor_t sap_minor;
        minor_t sap_lastminor;
        unsigned int sap_npush;
        char sap_list[MAXAPUSH][FMNAMSZ + 1];
};
\s+1\fP
.fi
.RE
.TP \w'sap_lastminor\(em'u
.I sap_cmd
This member is ignored on call an is unchanged on return.
.TP
.I sap_major
This member specifies the major device number of the
.IR STREAMS (4)
driver for which
to retrieve the current autopush list.
.TP
.I sap_minor
This member specifies the minor device number of the
.IR STREAMS (4)
driver for which
to retrieve the current autopush list.
.TP
.I sap_lastminor
This member is ignored on call an is unchanged on return.
.TP
.I sap_npush
Indicates on successful return the number of modules in the autopush list for
the
.IR STREAMS (4)
device specified in the
.IR sap_major " and " sap_minor
members.
Upon failure, this member remains unchanged.
.TP
.I sap_list
Indicates on successful return the null terminated names of the
.I sap_npush
modules in the autopush list for the
.IR STREAMS (4)
device specified in 
.IR sap_major " and " sap_minor .
Unused elements in the list are filled with character nulls.
Upon failure, this member remains unchanged.
.\"
.\"
.PP
For
.B SAD_GAP
to be successful,
.IR sap_major " and " sap_minor
must specify a
.IR STREAMS (4)
device that was previously configured for autopush with
.BR SAP_SAP ,
above,
or using
.BR register_strdev (9).
.\"
.\"
.PP
Upon success,
.B SAD_GAP
returns zero (0) and sets the
.IR sap_nlist " and " sap_list
members to indicate the current autopush settings for the
.IR STREAMS (4)
device
specified in
.IR sap_major " and " sap_minor .
Upon failure,
.B SAD_GAP
returns -1, does not alter any members, and sets
.IR errno (3)
to an appropriate error number as follows:
.TP \w'[EEXIST]\(em'u
.RB [ EFAULT ]
.I arg
is
.B NULL
or points outside the caller's valid address space.
.TP
.RB [ EINVAL ]
The device specified by
.IR sap_major " and " sap_minor
is invalid.
.TP
.RB [ ENOSTR ]
The device specified by
.IR sap_major " and " sap_minor
is not a
.IR STREAMS (4)
device.
.TP
.RB [ ENODEV ]
The device specified by
.IR sap_major " and " sap_minor
is not configured for autopush.
.PP
In addition, any of the errors returned by the
.BR ioctl (2)
command may be returned.
.\"
.\"
.SS SAD_VML
Verify that a module list is valid.
.PP
.I arg
points to a
.B str_list
structure formatted as follows:
.RS
.nf
\fC\s-1
struct str_list {
        int sl_nmods;
        struct str_mlist *sl_modlist;
};
\s+1\fP
.fi
.RE
.TP \w'sap_lastminor\(em'u
.I sl_nmods
.TP
.I sl_modlist
.RS
.nf
\fC\s-1
struct str_mlist {
        char l_name[FMNAMESZ + 1];
};
\s+1\fP
.fi
.RE
.\"
.\"
.PP
.B SAD_VML
has the side-effect that if a
.IR STREAMS (4)

NEWS
OpenSS7 Master Package -- history of user visible changes.  2008-05-03
$Id: NEWS,v 0.9.2.53 2008-05-03 21:22:25 brian Exp $
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2000  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
OpenSS7 Master Package package.

 Major changes for release openss7-0.9.2.G
------------------------------------------

This is the sixth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Documented `lib32gcc1' problem on Ubuntu.

   - Noted problem running under XEN kernels.  XEN kernels are not yet
     supported.

   - License upgrade to AGPL v3.

   - Cross-compile fixes (strap out AC_FUNC_REALLOC macro when
     cross-compiling).

   - Ability to strap out major documentation build and installation
     primarily for embedded targets.

   - Improvements to common build process for embedded and
     cross-compile targets.

   - Support for flex 2.5.33  in maintainer mode.

   - Added the following sub-packages to the master build (and relase):

         - `strx25-0.9.2.1'


_This is a public stable production grade release of the package: it
deprecates previous releases.  Please upgrade to the current release
before reporting bugs._

As with other OpenSS7 releases, this release configures, compiles,
installs and builds RPMs and DEBs for a wide range of Linux 2.4 and 2.6
RPM- and DPKG-based distributions, and can be used on production
kernels without patching or recompiling the kernel.

This package is publicly released under the `GNU Affero General Public
License Version 3'.  The release is available as an `autoconf' tarball,
SRPM, DSC, and set of binary RPMs and DEBs.  See the downloads page
(http://www.openss7.org/download.html) for the `autoconf' tarballs,
SRPMs and DSCs.  For tarballs, SRPMs, DSCs and binary RPMs and DEBs,
see the openss7 package page (http://www.openss7.org/openss7_pkg.html).

See `http://www.openss7.org/codefiles/openss7-0.9.2.G/ChangeLog' and
`http://www.openss7.org/codefiles/openss7-0.9.2.G/NEWS' in the release
for more information.  Also, see the `openss7.pdf' manual in the
release (also in html `http://www.openss7.org/openss7_manual.html').

 Major changes for release openss7-0.9.2.F
------------------------------------------

This is the fifth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

The `OpenSS7 Master Package' is not released as often as the
sub-packages.  As sub-packages are released more often, to rebuild the
master package with a new sub-package release, simply replace the
directory to which the sub-package belongs with the unpacked
sub-package release and then rebuild the master package.  This release
provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Support build on openSUSE 10.2.

   - Support build on Fedora 7 and 2.6.21 kernel.

   - Support build on CentOS 5.0 (RHEL5).

   - Support build on Ubuntu 7.04.

   - Updated to gettext 0.16.1.

   - Changes to support build on 2.6.20-1.2307.fc5 and
     2.6.20-1.2933.fc6 kernel.

   - Supports build on Fedora Core 6.

   - Support for recent distributions and tool chains.


 Major changes for release openss7-0.9.2.E
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  See
`README' in the release for a sub-package listing.  Most of the
sub-packages in the release are production grade for `Linux
Fast-STREAMS'.  All existing validation test suites run clean on
supported distributions and architectures.

It is unlikely that the `OpenSS7 Master Package' will be released as
frequently as before.  Sub-packages will be released more often.  To
rebuild the master package with a new sub-package release, simply
replace the directory to which the sub-package belongs with the
unpacked sub-package release and then rebuild the master package.  This
release provides support for recent distributions and tool chains.

Major features since the last public release are as follows:

   - Addition of the `osr61' sub-package that contains Dialogic(R) Open
     System Release 6.1 version 239 GPL drivers.

   - A few minor corrections to the common build process.

   - Support for autoconf 2.61, automake 1.10 and gettext 0.16.

   - Support for Ubuntu 6.10 distribution and bug fixes for i386
     kernels.

   - The package now looks for subpackages with a version number as
     unpacked by separate tarball.


 Major changes for release openss7-0.9.2.D
------------------------------------------

This is the fourth public release of the OpenSS7 Master Package.  The
sub-packages have been reorganized for this release.  See `README' in
the release for a sub-package listing.  Aside from sub-package
reorganization, the major difference from previous release is that this
release no longer contains `LiS'.  Too many of the sub-packages will
not even build against `LiS' because of its Stream head deficiencies.


LiS/drivers/str/sad.c
/*****************************************************************************

 @(#) $RCSfile: sad.c,v $ $Name:  $($Revision: 1.1.1.3.4.15 $) $Date: 2008-04-29 08:33:11 $

 -----------------------------------------------------------------------------

 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2000  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>, or write to the
 Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 -----------------------------------------------------------------------------

 Last Modified $Date: 2008-04-29 08:33:11 $ by $Author: brian $

 -----------------------------------------------------------------------------

 $Log: sad.c,v $
 Revision 1.1.1.3.4.15  2008-04-29 08:33:11  brian
 - update headers for Affero release

 Revision 1.1.1.3.4.14  2007/08/15 04:57:59  brian
 - GPLv3 updates

 *****************************************************************************/

#ident "@(#) $RCSfile: sad.c,v $ $Name:  $($Revision: 1.1.1.3.4.15 $) $Date: 2008-04-29 08:33:11 $"

static char const ident[] = "$RCSfile: sad.c,v $ $Name:  $($Revision: 1.1.1.3.4.15 $) $Date: 2008-04-29 08:33:11 $";

/*
 *  sad: STREAMS Administrative Driver
 *
 *  Version: 0.1
 *
 *  Copyright (C) 1999 Ole Husgaard (sparre@login.dknet.dk)
 *
 */

#include <sys/LiS/module.h>	/* must be VERY first include */

#include <sys/stream.h>

#include <sys/sad.h>

#include <sys/LiS/config.h>
#include <sys/osif.h>

/*
 *  Some configuration sanity checks
 */
#ifndef SAD__DRV
#error Not configured
#endif

#if !defined(SAD__CMAJORS) || SAD__CMAJORS != 1
#error There should be exactly one major number
#endif

#if !defined(SAD__CMAJOR_0)
#error The major number should be defined
#endif

#if !defined(SAD_N_MINOR)
#define SAD_N_MINOR 4
#endif

#ifndef STATIC
#define STATIC static
#endif
#ifndef INLINE
#define INLINE inline
#endif

STATIC int _RP sad_open(queue_t *, dev_t *, int, int, cred_t *);
STATIC int _RP sad_close(queue_t *, int, cred_t *);
STATIC int _RP sad_wput(queue_t *, mblk_t *);

STATIC struct module_info sad_minfo = {
	0,				/* Module ID number */
	"sad",				/* Module name */
	0,				/* Min packet size accepted */
	INFPSZ,				/* Max packet size accepted */
	0,				/* Hi water mark ignored */
	0				/* Low water mark ignored */
};

STATIC struct qinit sad_rinit = {
	NULL,				/* No read put */
	NULL,				/* No read service */
	sad_open,			/* Each open */
	sad_close,			/* Last close */
	NULL,				/* Reserved */
	&sad_minfo,			/* Information */
	NULL				/* No statistics */
};

STATIC struct qinit sad_winit = {
	sad_wput,			/* Write put */
	NULL,				/* No write service */
	NULL,				/* Ignored */
	NULL,				/* Ignored */
	NULL,				/* Reserved */
	&sad_minfo,			/* Information */
	NULL				/* No statistics */
};

#if !defined __NO_VERSION__
STATIC
#endif
struct streamtab sad_info = {
	&sad_rinit,			/* Read queue */
	&sad_winit,			/* Write queue */
	NULL,				/* Unused */
	NULL				/* Unused */
};

/*
 *  Private per-stream data
 */
struct priv {
	queue_t *rq;
	caddr_t uaddr;
	int ioc_state;
	str_list_t vml;
};

/*
 *  Values of priv->ioc_state
 */
#define ST_NONE		0
#define ST_SAP_IN	1
#define ST_GAP_IN	2
#define ST_GAP_OUT	3
#define ST_VML_IN1	4
#define ST_VML_IN2	5

/*
 *  Per stream storage
 */
STATIC struct priv sad_sad[SAD_N_MINOR];

/****************************************************************************/
/*                                                                          */
/*  copyin/copyout handling.                                                */
/*                                                                          */
/****************************************************************************/

STATIC void
sad_copyio(struct priv *p, mblk_t *mp, int type, caddr_t uaddr, size_t size)
{
	struct copyreq *req;

	mp->b_datap->db_type = type;
	mp->b_wptr = mp->b_rptr + sizeof(*req);
	LISASSERT(mp->b_wptr <= mp->b_datap->db_lim);

	req = (struct copyreq *) mp->b_rptr;
	req->cq_addr = uaddr;
	req->cq_size = size;
	req->cq_flag = 0;

	putnext(p->rq, mp);
}

STATIC void
sad_iocdata(struct priv *p, mblk_t *mp)
{
	struct copyresp *res = (struct copyresp *) mp->b_rptr;
	mblk_t *bp = mp->b_cont;
	str_list_t *sl;
	struct iocblk *iocp;
	int err = 0, ret = 0;

	if (res->cp_rval != 0) {
		err = (int) (-(intptr_t) (long) res->cp_rval);
		ret = -1;
		goto ioctl_done;
	}

	LISASSERT(bp != NULL);
	LISASSERT(bp->b_datap->db_type == M_DATA);

	switch (p->ioc_state) {
	case ST_SAP_IN:
		p->ioc_state = ST_NONE;
		LISASSERT(res->cp_cmd == SAD_SAP);
		if (res->cp_rval != 0) {
			err = -EFAULT;
			ret = -1;
		} else if (bp->b_wptr - bp->b_rptr == sizeof(struct strapush))
			err = lis_apush_set((struct strapush *) bp->b_rptr);
		else {
			err = -EINVAL;
			ret = -1;
		}

	      ioctl_done:iocp = (struct iocblk *) mp->b_rptr;
		mp->b_datap->db_type = (err < 0) ? M_IOCNAK : M_IOCACK;
		iocp->ioc_count = 0;
		iocp->ioc_error = -err;
		iocp->ioc_rval = ret;
		putnext(p->rq, mp);
		return;

	case ST_GAP_IN:
		p->ioc_state = ST_GAP_OUT;
		LISASSERT(res->cp_cmd == SAD_GAP);
		if (res->cp_rval != 0) {
			err = -EFAULT;
			ret = -1;
		} else if (bp->b_wptr - bp->b_rptr == sizeof(struct strapush))
			err = lis_apush_get((struct strapush *) bp->b_rptr);
		else {
			err = -EINVAL;
			ret = -1;
		}
		if (err < 0) {
			p->ioc_state = ST_NONE;
			goto ioctl_done;
		}
		sad_copyio(p, mp, M_COPYOUT, p->uaddr, sizeof(struct strapush));
		return;

	case ST_GAP_OUT:
		p->ioc_state = ST_NONE;
		LISASSERT(res->cp_cmd == SAD_GAP);
		if (res->cp_rval != 0) {
			err = -EFAULT;
			ret = -1;
		}
		goto ioctl_done;

	case ST_VML_IN1:
		LISASSERT(res->cp_cmd == SAD_VML);
		sl = (str_list_t *) bp->b_rptr;
		if (bp->b_wptr - bp->b_rptr != sizeof(str_list_t)
		    || (p->vml.sl_nmods = sl->sl_nmods) <= 0) {
			p->ioc_state = ST_NONE;
			err = -EINVAL;
			ret = -1;
			goto ioctl_done;
		}
		p->ioc_state = ST_VML_IN2;
		sad_copyio(p, mp, M_COPYIN, (caddr_t) sl->sl_modlist,
			   sl->sl_nmods * sizeof(str_mlist_t));
		return;
	case ST_VML_IN2:
		p->ioc_state = ST_NONE;
		LISASSERT(res->cp_cmd == SAD_VML);
		if (res->cp_rval != 0) {
			err = -EFAULT;
			ret = -1;
		} else if (bp->b_wptr - bp->b_rptr != p->vml.sl_nmods * sizeof(str_mlist_t)) {
			err = -EINVAL;
			ret = -1;
		}
		p->vml.sl_modlist = (struct str_mlist *) bp->b_rptr;
		ret = lis_apush_vml(&p->vml);
		err = 0;
		goto ioctl_done;

	default:
		err = -EINVAL;
		ret = -1;
		goto ioctl_done;
	}
}

/****************************************************************************/
/*                                                                          */
/*  ioctl handling.                                                         */
/*                                                                          */
/****************************************************************************/

STATIC INLINE void
sad_do_ioctl(struct priv *p, mblk_t *mp)
{
	struct iocblk *iocp;
	mblk_t *dp;
	int err;

	LISASSERT(p != NULL);

	LISASSERT(mp != NULL);
	LISASSERT(mp->b_datap->db_type == M_IOCTL);
	LISASSERT(mp->b_wptr - mp->b_rptr >= sizeof(struct iocblk));

	dp = mp->b_cont;
	LISASSERT(dp != NULL);
	LISASSERT(dp->b_datap->db_type == M_DATA);
	LISASSERT(mp->b_wptr - mp->b_rptr >= sizeof(void *));

	iocp = (struct iocblk *) mp->b_rptr;
	if (iocp->ioc_count != TRANSPARENT) {
		err = -EINVAL;
	      nak_it:mp->b_datap->db_type = M_IOCNAK;
		iocp->ioc_error = -err;
		putnext(p->rq, mp);
		return;
	}

	if (p->ioc_state != ST_NONE) {
		err = -EINVAL;
		goto nak_it;
	}

	p->uaddr = *(caddr_t *) dp->b_rptr;

	switch (iocp->ioc_cmd) {
	case SAD_SAP:
		if (iocp->ioc_uid != 0) {
			err = -EACCES;
			goto nak_it;
		}
		p->ioc_state = ST_SAP_IN;
		sad_copyio(p, mp, M_COPYIN, p->uaddr, sizeof(struct strapush));
		return;
	case SAD_GAP:
		p->ioc_state = ST_GAP_IN;
		sad_copyio(p, mp, M_COPYIN, p->uaddr, sizeof(struct strapush));
		return;
	case SAD_VML:
		p->ioc_state = ST_VML_IN1;
		sad_copyio(p, mp, M_COPYIN, p->uaddr, sizeof(str_list_t));
		return;
	default:
		err = -EINVAL;
		goto nak_it;
	}
}

/****************************************************************************/
/*                                                                          */
/*  The usual STREAMS driver stuff.                                         */
/*                                                                          */
/****************************************************************************/

STATIC int _RP
sad_open(queue_t *q, dev_t *devp, int flag, int sflag, cred_t *crp)
{
	dev_t i;

	if (sflag == CLONEOPEN) {
		for (i = 1; i < SAD_N_MINOR; i++)
			if (sad_sad[i].rq == NULL)
				break;
		if (i == SAD_N_MINOR)
			return ENXIO;
	} else {
		if ((i = getminor(*devp)) >= SAD_N_MINOR)
			return EBUSY;
	}
	*devp = makedevice(getmajor(*devp), i);

	q->q_ptr = WR(q)->q_ptr = &sad_sad[i];
	sad_sad[i].rq = q;
	sad_sad[i].ioc_state = ST_NONE;

	MODGET();

	return 0;
}

STATIC int _RP
sad_close(queue_t *q, int flag, cred_t *crp)
{
	struct priv *p = q->q_ptr;

	LISASSERT(p != NULL);

	p->rq = q->q_ptr = WR(q)->q_ptr = NULL;

	MODPUT();

	return 0;
}

STATIC int _RP
sad_wput(queue_t *q, mblk_t *mp)
{
	struct priv *p;

	LISASSERT(q != NULL);
	LISASSERT(mp != NULL);

	p = q->q_ptr;
	LISASSERT(p != NULL);

	switch (mp->b_datap->db_type) {
	case M_FLUSH:
		if (mp->b_rptr[0] & FLUSHW) {
			if (mp->b_rptr[0] & FLUSHBAND)
				flushband(q, mp->b_rptr[1], FLUSHDATA);
			else
				flushq(q, FLUSHDATA);
			mp->b_rptr[0] &= ~FLUSHW;
		}
		if (mp->b_rptr[0] & FLUSHR) {
			if (mp->b_rptr[0] & FLUSHBAND)
				flushband(RD(q), mp->b_rptr[1], FLUSHDATA);
			else
				flushq(RD(q), FLUSHDATA);
			qreply(q, mp);
		} else
			freemsg(mp);
		break;
	case M_IOCTL:
		sad_do_ioctl(p, mp);
		break;
	case M_IOCDATA:
		sad_iocdata(p, mp);
		break;
	default:
		freemsg(mp);
		break;
	}
	return (0);
}

/****************************************************************************/
/*                                                                          */
/*  Loadable module initialization and shutdown.                            */
/*                                                                          */
/****************************************************************************/

#if !defined __NO_VERSION__

#ifdef KERNEL_2_5
int
sad_init_module(void)
#else
int
init_module(void)
#endif
{
	int ret = lis_register_strdev(SAD__CMAJOR_0, &sad_info,
				      SAD_N_MINOR, LIS_OBJNAME_STR);

	if (ret < 0) {
		printk("%s: Unable to register module, error %d.\n", LIS_OBJNAME_STR, -ret);
		return ret;
	}
	memset(sad_sad, 0, sizeof(sad_sad));
	return 0;
}

#ifdef KERNEL_2_5
void
sad_cleanup_module(void)
#else
void
cleanup_module(void)
#endif
{
	int err = lis_unregister_strdev(SAD__CMAJOR_0);

	if (err < 0)
		printk("%s: Unable to unregister module, error %d.\n", LIS_OBJNAME_STR, -err);
	else
		printk("%s: Unregistered, ready to be unloaded.\n", LIS_OBJNAME_STR);
	return;
}

#ifdef KERNEL_2_5
module_init(sad_init_module);
module_exit(sad_cleanup_module);
#endif

#if defined(LINUX)		/* linux kernel */
#if defined(MODULE_LICENSE)
MODULE_LICENSE("GPL");
#endif
#if defined(MODULE_AUTHOR)
MODULE_AUTHOR("Ole Husgaard (sparre@login.dknet.dk");
#endif
#if defined(MODULE_DESCRIPTION)
MODULE_DESCRIPTION("STREAMS Administrative Driver");
#endif
#if defined(MODULE_ALIAS)
MODULE_ALIAS("streams-" __stringify(LIS_OBJNAME));
#endif
#endif

#endif				/* !defined __NO_VERSION__ */

#ifdef MODULE_ALIAS
MODULE_ALIAS("char-major-" __stringify(SAD__CMAJOR_0));
MODULE_ALIAS("char-major-" __stringify(SAD__CMAJOR_0) "-*");
MODULE_ALIAS("char-major-" __stringify(CLONE__CMAJOR_0) "-" __stringify(SAD__CMAJOR_0));
MODULE_ALIAS("/dev/sad");
#endif
